### PR TITLE
fix #20545 assign closure proc to const

### LINKED ISSUE #16736: `const` proc in `let` array causes invalid codegen
### Code
```nim
type ProcType = proc()
const something: ProcType = proc() = discard
let arr: array[1, ProcType] = [something]

```

### Current Output

```
/home/anuke/.cache/nim/bug_d/@mbug.nim.c: In function ‘NimMainModule’:
/home/anuke/.cache/nim/bug_d/@mbug.nim.c:413:99: error: request for member ‘ClE_0’ in something not a structure or union
  413 | lzADdspe4KtMwjNjbL8g[0].ClE_0), colonanonymous___DxAxlW4RjxKP1O7MGX2PnQ.ClE_0);
      |                                                                        ^

/home/anuke/.cache/nim/bug_d/@mbug.nim.c:414:80: error: request for member ‘ClP_0’ in something not a structure or union
  414 | tMwjNjbL8g[0].ClP_0 = colonanonymous___DxAxlW4RjxKP1O7MGX2PnQ.ClP_0;
      |                                                              ^

Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I'/home/anuke/.choosenim/toolchains/nim-#devel/lib' -I/home/anuke -o /home/anuke/.cache/nim/bug_d/@mbug.nim.c.o /home/anuke/.cache/nim/bug_d/@mbug.nim.c' failed with exit code: 1

```

### Expected Output
```
successful compilation, or at least a nim compilation error
```

### Workarounds

- Declaring both the proc and the array as `const` fixes the problem.

### Additional Information

- Occurs in 1.4.2 as well as latest HEAD
- Occurs with and without `--gc:arc`

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-01-16
Copyright (c) 2006-2021 by Andreas Rumpf

44ceefa9fefca58ba32e9638f754d7f4ca8a2b82 (current git head)
```

### LINKED ISSUE #20545: Calling a const proc produces an error message
### What happened?

```
type
  F = proc()

const
  f: F = proc() = echo "Got into f."

proc g() = echo "Got into g."

# Produces this compile-time error message:
#     Error: VM problem: dest register is not set
f()

# Just for contrast: if you comment out the line above, the code compiles
# and this next line runs just fine.
g()
```

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-10-10
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 1f2075042b7feb06cffe6c4ae6562a2e38144f81
active boot switches: -d:release

### Current Standard Output Logs

_No response_

### Expected Standard Output Logs

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bung87**
fix https://github.com/nim-lang/Nim/issues/20545
after this, below issues are more specific to array construction or assignment.  
may introduce new flag for later check.  

refs: https://github.com/nim-lang/Nim/issues/16736
refs: https://github.com/nim-lang/Nim/issues/16737

### CODE REVIEW

File: compiler/sem.nim [hunk 1]
```diff
@@ -375,6 +375,11 @@ proc semConstExpr(c: PContext, n: PNode; expectedType: PType = nil): PNode =
     return n
   if e.kind in nkSymChoices and e[0].typ.skipTypes(abstractInst).kind == tyEnum:
     return e
+  if e.typ.kind == tyProc and e.typ.callConv == ccClosure and {tfHasAsgn} * e.typ.flags != {}:
+    localError(c.config, n.info, "const closure proc with an environment cannot be used at runtime: " &
+                   renderTree(n, {renderNoComments}))
+    return e
+
```
Comment (Araq) [inline compiler/sem.nim:382]: Instead you need to inspect the **value**! `e[2].kind` must be `nkNilLit` or similar.

Comment (bung87) [reply compiler/sem.nim:382]: I dont understand, neither `e` or `n` length greater than 2 in the test case.

Comment (Araq) [reply compiler/sem.nim:382]: Sorry, I meant `e[1].kind`

Comment (bung87) [reply compiler/sem.nim:382]:
```
nkIntLit
cast[CSighandlerT](0)
nkEmpty
proc () {.closure.} = echo ["Hello"]
nkSym
makeMyClosure(myRef)
```
it's I get from test case.

**Comment by bung87** [comment]
succed by https://github.com/nim-lang/Nim/pull/20616

**Pull request closed without merging**